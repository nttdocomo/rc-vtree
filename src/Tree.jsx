import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import TreeNode from './TreeNode';
import { treeContextTypes } from './contextTypes';
import { CellMeasurer, CellMeasurerCache, AutoSizer, List } from 'react-virtualized';
import 'react-virtualized/styles.css';

import {
  nodeHasChildren,
  isNodeExpanded,
  getFlattenedTree,
  convertTreeToEntities,
  parseCheckedKeys,
  conductExpandParent,
  calcSelectedKeys,
  conductCheck,
  arrDel,
  arrAdd
} from './util';

const cache = new CellMeasurerCache({
  defaultHeight: 20,
  fixedWidth: true,
  keyMapper: () => 1,
});

export class TreeList extends React.Component {
  static propTypes = {
    prefixCls: PropTypes.string,
    defaultExpandParent: PropTypes.bool,
    autoExpandParent: PropTypes.bool,
    expandedKeys: PropTypes.arrayOf(PropTypes.string),
    defaultExpandedKeys: PropTypes.arrayOf(PropTypes.string),
    selectable: PropTypes.bool,
    selectedKeys: PropTypes.arrayOf(PropTypes.string),
    onExpand: PropTypes.func,
    switcherIcon: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
    onRightClick: PropTypes.func,
  }
  static defaultProps = {
    prefixCls: 'rc-tree',
    selectable: true,
    defaultExpandParent: true,
    autoExpandParent: false,
    defaultExpandedKeys: [],
  };
  static childContextTypes = treeContextTypes;

  constructor(props) {
    super(props);

    this.state = {
      keyEntities: {},
      flatTreeData: [],
      selectedKeys: [],
      checkedKeys: [],
      halfCheckedKeys: [],
    };
  
    // Internal usage for `rc-tree-select`, we don't promise it will not change.
    this.domTreeNodes = {};
  }

  getChildContext() {
    const {
      prefixCls, selectable, checkable, switcherIcon
    } = this.props;

    return {
      rcTree: {
        // root: this,

        prefixCls,
        checkable,
        selectable,
        switcherIcon,
        onNodeExpand: this.onNodeExpand,
        onNodeCheck: this.onNodeCheck,
        onNodeClick: this.onNodeClick,
        onNodeSelect: this.onNodeSelect,
        onNodeMouseEnter: this.onNodeMouseEnter,
        onNodeMouseLeave: this.onNodeMouseLeave,
        onNodeContextMenu: this.onNodeContextMenu,
      },
    };
  }
  getFlattenedTree = (nodes, parents = []) => {
    const {
      prefixCls, className, focusable,
      showLine, tabIndex = 0, treeData, defaultExpandAll
    } = this.props;
    return nodes.reduce((flattenedTree, { children, ...node }) => {
      const deepness = parents.length;
      const nodeWithHelpers = {...node, deepness, parents};

      if (!nodeHasChildren(children) || !isNodeExpanded(node)) {
        return [...flattenedTree, nodeWithHelpers];
      }

      return [...flattenedTree, nodeWithHelpers, ...getFlattenedTree(children, [...parents, node.id])];
    }, []);
  }
  static getDerivedStateFromProps(props, prevState) {
    const { prevProps } = prevState;
    const newState = {
      prevProps: props,
    };
    function needSync(name) {
      return (!prevProps && name in props) || (prevProps && prevProps[name] !== props[name]);
    }
    let flatTreeData = null;

    // Tree support filter function which will break the tree structure in the vdm.
    // We cache the treeNodes in state so that we can return the treeNode in event trigger.
    if (props.treeData) {
      newState.treeData = props.treeData;

      // Calculate the entities data for quick match
      const entitiesMap = convertTreeToEntities(props.treeData);
      newState.posEntities = entitiesMap.posEntities;
      newState.keyEntities = entitiesMap.keyEntities;
    }

    /*keyEntities
    key:{
      children:
      index:
      key:
      node:
      parent:
      pos:
    }
    */

    const keyEntities = newState.keyEntities || prevState.keyEntities;

    // ================ expandedKeys =================
    if (needSync('expandedKeys') || (prevProps && needSync('autoExpandParent'))) {
      newState.expandedKeys = (props.autoExpandParent || (!prevProps && props.defaultExpandParent)) ?
        conductExpandParent(props.expandedKeys, keyEntities) : props.expandedKeys;
    } else if (!prevProps && props.defaultExpandAll) {
      newState.expandedKeys = Object.keys(keyEntities);
    } else if (!prevProps && props.defaultExpandedKeys) {
      newState.expandedKeys = (props.autoExpandParent || props.defaultExpandParent) ?
        conductExpandParent(props.defaultExpandedKeys, keyEntities) : props.defaultExpandedKeys;
    }

    // ================ selectedKeys =================
    if (props.selectable) {
      if (needSync('selectedKeys')) {
        newState.selectedKeys = calcSelectedKeys(props.selectedKeys, props);
      } else if (!prevProps && props.defaultSelectedKeys) {
        newState.selectedKeys = calcSelectedKeys(props.defaultSelectedKeys, props);
      }
    }
    // Check if `treeData` or `children` changed and save into the state.
    /*if (needSync('treeData')) {
      flatTreeData = getFlattenedTree(props.treeData, [], newState.expandedKeys);
      newState.flatTreeData = flatTreeData;
    }*/
    // ================= checkedKeys =================
    if (props.checkable) {
      let checkedKeyEntity;

      if (needSync('checkedKeys')) {
        checkedKeyEntity = parseCheckedKeys(props.checkedKeys) || {};
      } else if (!prevProps && props.defaultCheckedKeys) {
        checkedKeyEntity = parseCheckedKeys(props.defaultCheckedKeys) || {};
      }
      if (checkedKeyEntity) {
        let { checkedKeys = [], halfCheckedKeys = [] } = checkedKeyEntity;

        if (!props.checkStrictly) {
          const conductKeys = conductCheck(checkedKeys, true, keyEntities);
          checkedKeys = conductKeys.checkedKeys;
          halfCheckedKeys = conductKeys.halfCheckedKeys;
        }

        newState.checkedKeys = checkedKeys;
        newState.halfCheckedKeys = halfCheckedKeys;
      }
    }
    return newState;
  }

  onNodeClick = (e, treeNode) => {
    const { onClick } = this.props;
    if (onClick) {
      onClick(e, treeNode);
    }
  };


  onNodeSelect = (e, treeNode) => {
    let { selectedKeys } = this.state;
    const { keyEntities } = this.state;
    const { onSelect, multiple } = this.props;
    const { selected, eventKey } = treeNode.props;
    const targetSelected = !selected;

    // Update selected keys
    if (!targetSelected) {
      selectedKeys = arrDel(selectedKeys, eventKey);
    } else if (!multiple) {
      selectedKeys = [eventKey];
    } else {
      selectedKeys = arrAdd(selectedKeys, eventKey);
    }

    // [Legacy] Not found related usage in doc or upper libs
    const selectedNodes = selectedKeys.map(key => {
      const entity = keyEntities[key];
      if (!entity) return null;

      return entity.node;
    }).filter(node => node);

    this.setUncontrolledState({ selectedKeys });

    if (onSelect) {
      const eventObj = {
        event: 'select',
        selected: targetSelected,
        node: treeNode,
        selectedNodes,
        nativeEvent: e.nativeEvent,
      };
      onSelect(selectedKeys, eventObj);
    }
  };
  onNodeCheck = (e, treeNode, checked) => {
    const { keyEntities, checkedKeys: oriCheckedKeys, halfCheckedKeys: oriHalfCheckedKeys } = this.state;
    const { checkStrictly, onCheck } = this.props;
    const { props: { eventKey } } = treeNode;

    // Prepare trigger arguments
    let checkedObj;
    const eventObj = {
      event: 'check',
      node: treeNode,
      checked,
      nativeEvent: e.nativeEvent,
    };

    if (checkStrictly) {
      const checkedKeys = checked ? arrAdd(oriCheckedKeys, eventKey) : arrDel(oriCheckedKeys, eventKey);
      const halfCheckedKeys = arrDel(oriHalfCheckedKeys, eventKey);
      checkedObj = { checked: checkedKeys, halfChecked: halfCheckedKeys };

      eventObj.checkedNodes = checkedKeys
        .map(key => keyEntities[key])
        .filter(entity => entity)
        .map(entity => entity.node);

      this.setUncontrolledState({ checkedKeys });
    } else {
      const { checkedKeys, halfCheckedKeys } = conductCheck([eventKey], checked, keyEntities, {
        checkedKeys: oriCheckedKeys, halfCheckedKeys: oriHalfCheckedKeys,
      });

      checkedObj = checkedKeys;

      // [Legacy] This is used for `rc-tree-select`
      /*eventObj.checkedNodes = [];*/
      eventObj.checkedNodesPositions = [];
      eventObj.halfCheckedKeys = halfCheckedKeys;

      checkedKeys.forEach((key) => {
        const entity = keyEntities[key];
        if (!entity) return;

        const { pos } = entity;

        //eventObj.checkedNodes.push(node);
        eventObj.checkedNodesPositions.push({ pos });
      });

      this.setUncontrolledState({
        checkedKeys,
        halfCheckedKeys,
      });
    }

    if (onCheck) {
      onCheck(checkedObj, eventObj);
    }
  }
  onNodeExpand = (e, treeNode) => {
    let { expandedKeys } = this.state;
    const { onExpand } = this.props;
    const { eventKey, expanded } = treeNode.props;
    const targetExpanded = !expanded;
    if (targetExpanded) {
      expandedKeys = arrAdd(expandedKeys, eventKey);
    } else {
      expandedKeys = arrDel(expandedKeys, eventKey);
    }
    this.setUncontrolledState({ expandedKeys });
    if (onExpand) {
      onExpand(expandedKeys, {
        node: treeNode,
        expanded: targetExpanded,
        nativeEvent: e.nativeEvent,
      });
    }
  };

  onNodeMouseEnter = (event, node) => {
    const { onMouseEnter } = this.props;
    if (onMouseEnter) {
      onMouseEnter({ event, node });
    }
  };

  onNodeMouseLeave = (event, node) => {
    const { onMouseLeave } = this.props;
    if (onMouseLeave) {
      onMouseLeave({ event, node });
    }
  };

  onNodeContextMenu = (event, node) => {
    const { onRightClick } = this.props;
    if (onRightClick) {
      event.preventDefault();
      onRightClick({ event, node });
    }
  };
  /**
   * Only update the value which is not in props
   */
  setUncontrolledState = (state) => {
    let needSync = false;
    const newState = {};

    Object.keys(state).forEach(name => {
      if (name in this.props) return;

      needSync = true;
      newState[name] = state[name];
    });

    if (needSync) {
      this.setState(newState);
    }
  };
  isKeyChecked = (key) => {
    const { checkedKeys = [] } = this.state;
    return checkedKeys.indexOf(key) !== -1;
  };
  renderTreeNode = (item, {index, isScrolling, key, parent, style}) => {
    const {
      rowRenderer
    } = this.props;
    const {
      expandedKeys = [],
      halfCheckedKeys = [],
      selectedKeys = []
    } = this.state;
    const { children, isLeaf, title, ...props } = item
    let child

    if(rowRenderer){
      const { children, isLeaf, title, ...props } = item
      child = rowRenderer(item, {index, isScrolling, key, parent, style})
      return React.cloneElement(child, {
        isLeaf: isLeaf,
        eventKey: item.key,
        expanded: expandedKeys.indexOf(item.key) !== -1,
        selected: selectedKeys.indexOf(item.key) !== -1,
        checked: this.isKeyChecked(item.key),
        halfChecked: halfCheckedKeys.indexOf(item.key) !== -1,
        style:style,
        ...props
      });
    } else {
      const { children, isLeaf, ...props } = item
      return <TreeNode
        {...props}
        isLeaf={isLeaf}
        eventKey={item.key}
        expanded={expandedKeys.indexOf(item.key) !== -1}
        checked={this.isKeyChecked(item.key)}
        selected={selectedKeys.indexOf(item.key) !== -1}
        halfChecked={halfCheckedKeys.indexOf(item.key) !== -1}
        style={style}
      />
    }
  }
  render(){
    const {
      prefixCls, className, focusable,
      showLine, tabIndex = 0, treeData,
      height, width, rowRenderer
    } = this.props;
    const {
      expandedKeys = [],
      halfCheckedKeys = [],
      selectedKeys = []
    } = this.state;
    const flatTreeData = getFlattenedTree(treeData, [], expandedKeys);
    return <List
      className={classNames(prefixCls, className, {
        [`${prefixCls}-show-line`]: showLine,
      })}
      role="tree"
      unselectable="on"
      height={height}
      overscanRowCount={1}
      //ref:setRef,
      rowHeight={cache.rowHeight}
      rowRenderer={({index, isScrolling, key, parent, style}) => {
        const item = flatTreeData[index]
        if(!item){
          return null
        }
        //return convertDataToTree(item)
        return <CellMeasurer
          cache={cache}
          columnIndex={0}
          key={key}
          parent={parent}
          rowIndex={index}
        >
          {this.renderTreeNode(item, {index, isScrolling, key, parent, style})}
          {/*
          <TreeNode
            {...props}
            isLeaf={item.isLeaf}
            eventKey={item.key}
            expanded={expandedKeys.indexOf(item.key) !== -1}
            checked={this.isKeyChecked(item.key)}
            selected={selectedKeys.indexOf(item.key) !== -1}
            halfChecked={halfCheckedKeys.indexOf(item.key) !== -1}
            style={style}
          />*/}
        </CellMeasurer>;
        // return this.renderTreeNode(treeNode, index)
        /*return React.cloneElement(child, {
          style:Object.assign({}, style, {
            paddingLeft: 18*child.props.pos.split('-')[0]
          })
        });*/
        //return this.renderTreeNode(nodeList[index], index, style)
        /*return mapChildren(nodeList, (node, index) => (
          this.renderTreeNode(node, index)
        ));*/
      }}
      onRowsRendered={this.props.onRowsRendered}
      rowCount={flatTreeData.length}
      width={width}
    />
  }
}

export function Tree (props) {
  return (
    <AutoSizer>
      {(param) => {
        const newProps = {...props, ...param}
        return <TreeList
          {...newProps}
        />
      }}
    </AutoSizer>
  );
}

export default Tree;