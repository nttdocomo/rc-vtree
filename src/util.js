import React, { Children } from 'react';
import toArray from 'rc-util/lib/Children/toArray';
import TreeNode from './VTreeNode';
import warning from 'warning';

export const isNodeExpanded = (node, expandedKeys) => expandedKeys.indexOf(node.key) > -1;
export const nodeHasChildren = children => children !== undefined && children.length > 0;

export function arrDel(list, value) {
  const clone = list.slice();
  const index = clone.indexOf(value);
  if (index >= 0) {
    clone.splice(index, 1);
  }
  return clone;
}

export function arrAdd(list, value) {
  const clone = list.slice();
  if (clone.indexOf(value) === -1) {
    clone.push(value);
  }
  return clone;
}

export function isCheckDisabled(node) {
  const { disabled, disableCheckbox } = node.props || {};
  return !!(disabled || disableCheckbox);
}

const internalProcessProps = props => props;
export function convertDataToTree(treeData, processer) {
  if (!treeData) return [];

  const { processProps = internalProcessProps } = processer || {};
  const list = Array.isArray(treeData) ? treeData : [treeData];
  return list.map(({ children, ...props }) => {
    return (
      <TreeNode {...processProps(props)}>
      </TreeNode>
    );
  });
}

export const getFlattenedTree = (nodes, parents = [], expandedKeys = []) =>
  nodes.reduce((flattenedTree, { children, ...node }) => {
    const deepness = parents.length;
    const nodeWithHelpers = {...node, deepness, parents, isLeaf:!nodeHasChildren(children)};

    if (!nodeHasChildren(children) || !isNodeExpanded(node, expandedKeys)) {
      return [...flattenedTree, nodeWithHelpers];
    }

    return [...flattenedTree, nodeWithHelpers, ...getFlattenedTree(children, [...parents, node.key], expandedKeys)];
  }, []);

export function getPosition(level, index) {
  return `${level}-${index}`;
}

export function getNodeChildren(children) {
  return toArray(children).filter(isTreeNode);
}

export function traverseTreeNodes(treeNodes, callback) {
  function processNode(node, index, parent) {
    const children = node ? node.children : treeNodes;
    const pos = node ? getPosition(parent.pos, index) : 0;

    // Filter children
    //const childList = getNodeChildren(children);

    // Process node if is not root
    if (node) {
      const data = {
        node,
        index,
        pos,
        key: node.key || pos,
        parentPos: parent.node ? parent.pos : null,
      };

      callback(data);
    }
    if(children){
      children.forEach((subNode, subIndex)=>{
        processNode(subNode, subIndex, { node, pos });
      })
    }

    // Process children node
    /*Children.forEach(childList, (subNode, subIndex) => {
      processNode(subNode, subIndex, { node, pos });
    });*/
  }

  processNode(null);
}

// TODO: ========================= NEW LOGIC =========================
/**
 * Calculate treeNodes entities. `processTreeEntity` is used for `rc-tree-select`
 * @param treeNodes
 * @param processTreeEntity  User can customize the entity
 */
export function convertTreeToEntities(treeNodes) {


  const posEntities = {};
  const keyEntities = {};
  let wrapper = {
    posEntities,
    keyEntities,
  };

  traverseTreeNodes(treeNodes, (item) => {
    const { index, pos, key, parentPos } = item;
    const entity = { index, key, pos };

    posEntities[pos] = entity;
    keyEntities[key] = entity;

    // Fill children
    entity.parent = posEntities[parentPos];
    if (entity.parent) {
      entity.parent.children = entity.parent.children || [];
      entity.parent.children.push(entity);
    }
  });

  /*if (onProcessFinished) {
    onProcessFinished(wrapper);
  }*/

  return wrapper;
}

/**
 * Return selectedKeys according with multiple prop
 * @param selectedKeys
 * @param props
 * @returns [string]
 */
export function calcSelectedKeys(selectedKeys, props) {
  if (!selectedKeys) return undefined;

  const { multiple } = props;
  if (multiple) {
    return selectedKeys.slice();
  }

  if (selectedKeys.length) {
    return [selectedKeys[0]];
  }
  return selectedKeys;
}



/**
 * Since React internal will convert key to string,
 * we need do this to avoid `checkStrictly` use number match
 */
function keyListToString(keyList) {
  if (!keyList) return keyList;
  return keyList.map(key => String(key));
}

/**
 * Parse `checkedKeys` to { checkedKeys, halfCheckedKeys } style
 */
export function parseCheckedKeys(keys) {
  if (!keys) {
    return null;
  }

  // Convert keys to object format
  let keyProps;
  if (Array.isArray(keys)) {
    // [Legacy] Follow the api doc
    keyProps = {
      checkedKeys: keys,
      halfCheckedKeys: undefined,
    };
  } else if (typeof keys === 'object') {
    keyProps = {
      checkedKeys: keys.checked || undefined,
      halfCheckedKeys: keys.halfChecked || undefined,
    };
  } else {
    warning(false, '`checkedKeys` is not an array or an object');
    return null;
  }

  keyProps.checkedKeys = keyListToString(keyProps.checkedKeys);
  keyProps.halfCheckedKeys = keyListToString(keyProps.halfCheckedKeys);

  return keyProps;
}

/**
 * Conduct check state by the keyList. It will conduct up & from the provided key.
 * If the conduct path reach the disabled or already checked / unchecked node will stop conduct.
 * @param keyList       list of keys
 * @param isCheck       is check the node or not
 * @param keyEntities   parsed by `convertTreeToEntities` function in Tree
 * @param checkStatus   Can pass current checked status for process (usually for uncheck operation)
 * @returns {{checkedKeys: [], halfCheckedKeys: []}}
 */
export function conductCheck(keyList, isCheck, keyEntities, checkStatus = {}) {
  const checkedKeys = {};
  const halfCheckedKeys = {}; // Record the key has some child checked (include child half checked)

  (checkStatus.checkedKeys || []).forEach((key) => {
    checkedKeys[key] = true;
  });

  (checkStatus.halfCheckedKeys || []).forEach((key) => {
    halfCheckedKeys[key] = true;
  });
  // Conduct up
  function conductUp(key) {
    if (checkedKeys[key] === isCheck) return;

    const entity = keyEntities[key];
    if (!entity) return;

    const { children, parent, ...node } = entity;

    if (isCheckDisabled(node)) return;

    // Check child node checked status
    let everyChildChecked = true;
    let someChildChecked = false; // Child checked or half checked

    (children || [])
      .filter(child => !isCheckDisabled(child))
      .forEach(({ key: childKey }) => {
        const childChecked = checkedKeys[childKey];
        const childHalfChecked = halfCheckedKeys[childKey];

        if (childChecked || childHalfChecked) someChildChecked = true;
        if (!childChecked) everyChildChecked = false;
      });

    // Update checked status
    if (isCheck) {
      checkedKeys[key] = everyChildChecked;
    } else {
      checkedKeys[key] = false;
    }
    halfCheckedKeys[key] = someChildChecked;

    if (parent) {
      conductUp(parent.key);
    }
  }

  // Conduct down
  function conductDown(key) {
    if (checkedKeys[key] === isCheck) return;

    const entity = keyEntities[key];
    if (!entity) return;

    const { children } = entity;

    if (isCheckDisabled(entity)) return;

    checkedKeys[key] = isCheck;

    (children || []).forEach((child) => {
      conductDown(child.key);
    });
  }

  function conduct(key) {
    const entity = keyEntities[key];

    if (!entity) {
      warning(false, `'${key}' does not exist in the tree.`);
      return;
    }

    const { children, parent } = entity;
    checkedKeys[key] = isCheck;

    if (isCheckDisabled(entity)) return;

    // Conduct down
    (children || [])
      .filter(child => !isCheckDisabled(child))
      .forEach((child) => {
        conductDown(child.key);
      });

    // Conduct up
    if (parent) {
      conductUp(parent.key);
    }
  }

  (keyList || []).forEach((key) => {
    conduct(key);
  });

  const checkedKeyList = [];
  const halfCheckedKeyList = [];

  // Fill checked list
  Object.keys(checkedKeys).forEach((key) => {
    if (checkedKeys[key]) {
      checkedKeyList.push(key);
    }
  });

  // Fill half checked list
  Object.keys(halfCheckedKeys).forEach((key) => {
    if (!checkedKeys[key] && halfCheckedKeys[key]) {
      halfCheckedKeyList.push(key);
    }
  });

  return {
    checkedKeys: checkedKeyList,
    halfCheckedKeys: halfCheckedKeyList,
  };
}

/**
 * If user use `autoExpandParent` we should get the list of parent node
 * @param keyList
 * @param keyEntities
 */
export function conductExpandParent(keyList, keyEntities) {
  const expandedKeys = {};

  function conductUp(key) {
    if (expandedKeys[key]) return;

    const entity = keyEntities[key];
    if (!entity) return;

    expandedKeys[key] = true;

    const { parent } = entity;

    if (isCheckDisabled(entity)) return;

    if (parent) {
      conductUp(parent.key);
    }
  }

  (keyList || []).forEach((key) => {
    conductUp(key);
  });

  return Object.keys(expandedKeys);
}