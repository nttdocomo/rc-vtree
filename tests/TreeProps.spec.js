import React from 'react';
import { render, mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import Animate from 'rc-animate';
import PropTypes from 'prop-types';
import Tree, { TreeNode } from '..';
import { nodeMatcher } from './util';

describe('Tree Props', () => {
  describe('custom switcher icon', () => {
    function switcherIcon(text, testLeaf) {
      const sfc = ({ isLeaf }) => {
        if (testLeaf) {
          return isLeaf ? <span>{text}</span> : null;
        }
        return isLeaf ? null : <span>{text}</span>;
      };

      sfc.propTypes = {
        isLeaf: PropTypes.bool,
      };

      return sfc;
    }
    it('switcher icon', () => {
      const wrapper = render(
        <Tree
          defaultExpandAll
          switcherIcon={switcherIcon('switcherIcon')}
          treeData={[{
          	key:"0-0"
          }, {
          	key:"0-1",
          	children: [{
          	  key:"0-1-0"
          	}, {
          	  key:"0-1-1"
          	}]
          }]}
        >
        </Tree>
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
})