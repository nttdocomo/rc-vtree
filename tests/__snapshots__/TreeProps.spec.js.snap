// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Props custom switcher icon switcher icon 1`] = `
<div
  style="overflow:visible;height:0;width:0"
>
  <div
    aria-label="grid"
    aria-readonly="true"
    class="ReactVirtualized__Grid ReactVirtualized__List rc-tree"
    role="tree"
    style="box-sizing:border-box;direction:ltr;height:0;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:auto"
    tabindex="0"
  />
</div>
`;
